addons/stock/models/stock_inventory.py

```diff
@@ -467,3 +528,31 @@ def _generate_moves(self):
                 vals = line._get_move_values(abs(diff), line.location_id.id, line.product_id.property_stock_inventory.id, True)
             vals_list.append(vals)
         return self.env['stock.move'].create(vals_list)
+
+    def action_refresh_quantity(self):
+        for line in self:
+            if line.outdated:
+                quant = self.env['stock.quant'].search([
+                    ('product_id', '=', line.product_id.id),
+                    ('location_id', '=', line.location_id.id),
+                ], limit=1)
+                if quant.exists() and not line.theoretical_qty == quant.quantity:
+                    line.theoretical_qty = quant.quantity
+                line.inventory_date = fields.Datetime.now()
+                line.outdated = False
+
+    def action_reset_product_qty(self):
+        self.update({'product_qty': 0})
+
+    @api.model
+    def action_validate_inventory(self, inventory_id):
+        if not self.user_has_groups('stock.group_stock_manager'):
+            return 'access_denied'
```

Comment:
why not raising a user error 'Only a stock manager can validate an inventory adjustment.'